src/cookbooks/bigquery/optimization.md

```diff
@@ -1,28 +1,165 @@
+different keys. In that case, data from both tables needs to be shuffled to every system.
+
+#### Key takeaways
+
+- Use `LIMIT` for query prototyping to dramatically reduce the volume of data scanned
+  as well as speeding up processing.
+- Use approximate algorithms. Then less data needs to be shuffled because
+  probabilistic data structures can be used instead of the raw data itself.
+- Specify large tables first in a `JOIN` operation. In this case, small tables can be sent to
+  every system to eliminate one data shuffle operation. Note that Spark supports a `broadcast`
+  command explicitly.
+
+### How is the data stored?
+
+The data is stored in columnar format.
+
+#### Traditional Row Stores
+
+Consider a typical CSV file, which represents an example of a row store.
+
+```
+name,age,height
+"Ted",27,6.0
+"Emmanuel",45,5.9
+"Cadence",5,3.5
+```
+
+When this data is stored to disk, you can read an entire record in consecutive order. For example, if
+the first `"` is stored at block 1 on disk, then a sequential scan from 1 assigns the first row of
+data: `"ted",27,6.0`. Keep scanning and you get `\n"Emm`... and so on.
+
+You can use the following query that you can execute very quickly:
+
+```
+SELECT *
+FROM people
+WHERE name == 'Ted'
+```
+
+The database can just scan the first row of data. However, the following is more difficult:
+
+```
+SELECT name
+FROM people
+```
+
+Now the database has to read _all_ of the rows and then select the `name` column, which results in a lot
+more overhead.
+
+#### Columnar Stores
+
+Columnar turns the data sideways. For example, you can make a columnar version of the above data
+and still store it in CSV format:
+
+```
+name,"Ted","Emmanuel","Cadence"
+age,27,45,5
+height,6.0,5.9,3.5
+```
+
+Now let's consider how we can query the data when it's stored this way:
+
+```
+SELECT *
+FROM people
+WHERE name == "ted"
+```
+
+In this case, all the data must be read because the
+`(name, age, height)` is not stored together.
+
+Here's another query:
+
+```
+SELECT name
+FROM people
+```
+
+In this case, only the "name" row needs to be read. All the other lines of the
+file can be skipped.
+
+#### Data partitions
+
+You can improve performance even further by taking advantage of partitions. These are entire files of data that share a value for a column. For example, if everyone in the `people` table lived in `DE`, then you can add that to the filename: `/country=DE/people.csv`.
+
+From there, a query engine would have to know how to read that path and understand that
+all of these people share a country. You can query as follows:
+
+```
+SELECT *
+FROM people
+WHERE country == 'US'
+```
+
+In this case, the query engine no longer even has to read the file. It could just look at the path and realize that there is nothing of interest.
+
+Tables are usually partitioned based on dates; e.g., `submission_date` or `DATE(submission_timestamp)`.
+
```

Comment:
```suggestion

#### Data Ordering (clustering)

Another way to improve query performance is to select a subset of data on a field that the data is ordered by. In BigQuery, this is called "clustering". A clustered field is one which is sorted in the underlying data.

For example, if we wanted to get all of ages greater than age 40 in our table above, we might query like this:
`SELECT age FROM people WHERE age > 45`

Then we would scan all of the `age` field, starting from `27`, then `45`, then `5`.

However, if we sort the data on that field, our table would look like this:
\```
name,"Cadence","Ted","Emmanuel"
age,5,27,45
height,3.5,6.0,5.9
\```

Since data is stored on different files, we could ignore any `age` files that don't have data less than 40. So if Cadence and Ted's ages were in one file, and Emmanuel's in the next, we could skip reading that first file entirely. In that way, we can sometimes drastically reduce the amount of data we're reading.
```

Reply:
LGTM, thanks!

Reply:
From:

For example, if we wanted to get all of ages greater than age 40 in our table above, we might query like this:

To:

For example, if you want to retrieve information in the table above-listed table about all age groups with members who are older than 40 years, you may wan to query as follows:

Reply:
I am sorry that there are typos in the above recommendation. Here are the fixes:
For example, if you want to retrieve information in the above-listed table about all age groups with members who are older than 40 years, you may want to query as follows:

Reply:
From:
For example, if we wanted to get all of ages greater than age 40 in our table above, we might query like this:

To:
For example, if you want to retrieve information in the above-listed table about all age groups with members who are older than 40 years, you may want to query as follows:

Reply:
From:
Then we would scan all of the `age` field, 

To:
Then you can scan all of the `age` field,

Reply:
From:
However, if we sort the data on that field, our table would look like this:

To:
If you sort data on that field, the table then looks as follows:

Reply:
From:
Since data is stored on different files, we could ignore any `age` files that don't have data less than 40. So if Cadence and Ted's ages were in one file, and Emmanuel's in the next, we could skip reading that first file entirely. In that way, we can sometimes drastically reduce the amount of data we're reading.

To:
You can ignore any `age` files that do not include any age-related data lower than 40 because data is stored in different files. If Cadence's and Ted's ages are included in one file and Emmanuel's age is included in the next file, you can then skip reading the first file. Consequently, the volume of data that is read in that manner can be drastically reduced.

Reply:
Side note from Elke: The word 'since' should only be used when indicating a particular point in time. For example, "Since I graduated from college, I have been working as a technical writer."

Most of the time 'since' is incorrectly used and should be replaced with the word 'because'. In other words, "As a cause of this and that occurring, I am no longer a technical writer." Many times we use in our verbal communication 'since' in this situation even though we should have said something like this "I am no longer a technical writer because (not since) I made so many mistakes. However, I stopped working since I got engaged (a particular point in time).